src/sources/apache_metrics/parser.rs

```diff
@@ -0,0 +1,852 @@
+use crate::event::metric::{Metric, MetricKind, MetricValue};
+use chrono::{DateTime, Utc};
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::error::Error;
+use std::fmt;
+
+/// Parses the text output from Apache's mod_status and returns:
+///
+/// - A list of metrics generated from the output
+/// - A list of parse errors that were encountered
+///
+/// # Arguments
+///
+/// - `payload` - the mod_status output
+/// - `namespace` - the namespace to put the generated metrics in
+/// - `now` - the time the payload was fetched
+/// - `tags` - any base tags to apply to the metrics
+pub fn parse(
+    payload: &str,
+    namespace: &str,
+    now: DateTime<Utc>,
+    tags: Option<&BTreeMap<String, String>>,
+) -> (Vec<Metric>, Vec<ParseError>) {
+    let mut parsed = payload
+        .lines()
+        .into_iter()
+        .filter_map(|l| {
+            let mut parts = l.splitn(2, ":");
+            let key = parts.next();
+            let value = parts.next().map(|s| s.trim());
+            match (key, value) {
+                (Some(k), Some(v)) => Some((k, v)),
+                _ => None,
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // mod_status has BusyWorkers/IdleWorkers repeated
+    // TODO better way to do this without .collect()ing? Do we care?
+    parsed.sort();
+    parsed.dedup();
+
+    parsed
+        .iter()
+        .map(|(key, value)| line_to_metrics(key, value, namespace, now, tags))
+        .fold(
+            (Vec::new(), Vec::new()),
+            |(mut metrics, mut errs), current| {
+                match current {
+                    LineResult::Metrics(m) => metrics.extend(m),
+                    LineResult::Error(err) => errs.push(err),
+                    LineResult::None => {}
+                }
+                (metrics, errs)
+            },
+        )
+}
+
+#[derive(Debug)]
+pub struct ParseError {
+    key: String,
+    err: Box<dyn Error>,
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "could not parse value for {}: {}", self.key, self.err)
+    }
+}
+
+impl Error for ParseError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(self.err.as_ref())
+    }
+}
+
+enum LineResult {
```

Comment:
Is this sort of `enum` idiomatic?

Reply:
This seems similar to `Option<Result<Vec<Metric>, ParseError>>`?

Reply:
:thinking: yeah it is basically that. I'll try swapping to remove the type.